### PR TITLE
add btime for linux

### PR DESCRIPTION
**Pull request by mateusz834**
It adds btime support for linux using statx syscall.

### CODE REVIEW

File: times_linux.go [hunk 1]
```diff
@@ -26,6 +30,87 @@ type timespec struct {
...
+	mtime
+	ctime
+	btime
+}
+
+// Stat returns the Timespec for the given filename.
+func Stat(name string) (Timespec, error) {
+	var statx unix.Statx_t
+
+	err := unix.Statx(unix.AT_FDCWD, name, unix.AT_EMPTY_PATH|unix.AT_STATX_SYNC_AS_STAT, unix.STATX_ATIME|unix.STATX_MTIME|unix.STATX_CTIME|unix.STATX_BTIME, &statx)
```
Comment (djherbis) [inline times_linux.go:44]: I usually like to link to the spec somewhere for these more complicated calls so readers know where all this comes from:
https://man7.org/linux/man-pages/man2/statx.2.html

Comment (djherbis) [inline times_linux.go:44]: Looking at [AT_EMPTY_PATH](https://man7.org/linux/man-pages/man2/statx.2.html#:~:text=the%0A%20%20%20%20%20%20%20following%20constants%3A-,AT_EMPTY_PATH,-If%20pathname%20is) does this mimic the behavior of the Stat command by default? Just curious.

File: times_linux.go [hunk 2]
```diff
@@ -26,6 +30,87 @@ type timespec struct {
...
+			fi, err := file.Stat()
+			if err != nil {
+				return nil, err
+			}
+			return getTimespec(fi), nil
+		}
+		return nil, err
+	}
+
+	runtime.KeepAlive(file)
```
[the remark this answers is not in the document]
Comment (mateusz834) [reply times_linux.go:89]: there is a runtime finalizer on *os.File, so I think it is required.

File: times_linux.go [hunk 3]
```diff
@@ -26,6 +30,87 @@ type timespec struct {
...
+}
+
+// Stat returns the Timespec for the given filename.
+func Stat(name string) (Timespec, error) {
+	var statx unix.Statx_t
+
+	err := unix.Statx(unix.AT_FDCWD, name, unix.AT_EMPTY_PATH|unix.AT_STATX_SYNC_AS_STAT, unix.STATX_ATIME|unix.STATX_MTIME|unix.STATX_CTIME|unix.STATX_BTIME, &statx)
+	if err != nil {
+		//linux 4.10 and earlier does not support Statx syscall
+		if errors.Is(err, unix.ENOSYS) {
```
[the remark this answers is not in the document]
Comment (mateusz834) [reply times_linux.go:47]: I tested it using kernel 4.9 and the fallback worked as expected.
ENOSYS -> “Function not implemented.” This indicates that the function called is not implemented at all, either in the C library itself or in the operating system. When you get this error, you can be sure that this particular function will always fail with ENOSYS unless you install a new version of the C library or the operating system. 
http://www.gnu.org/software/libc/manual/html_node/Error-Codes.html

**Comment by mateusz834** [comment]
I replaced the (*os.File).Fd() with (*os.File).SyscallConn(), because the Fd() sets the fd to blocking mode.

**Comment by gildesmarais** [comment]
I came here via the mentioned Navidrome PR. Thank you for maintaining this library and thank you for this PR. 👍🏽 

How are chances to see a merge of this PR? :)

**Comment by djherbis** [comment]
@gildesmarais Merged, I'll do some testing and maybe some minor refactoring before marking it as a release.

**Comment by certuna** [comment]
Cool - will that be v1.6.0?

**Comment by djherbis** [comment]
It should be, it still needs testing with Linux with btime available I believe to verify it works.

I haven't had a chance to try it yet.

**Comment by djherbis** [comment]
Released: https://github.com/djherbis/times/releases/tag/v1.6.0

**Comment by djherbis** [comment]
I verified that I did get btime in a linux codespace :)

**Pull request merged**